akka-remote/src/main/scala/akka/remote/MessageSerializer.scala

```diff
@@ -53,21 +58,27 @@ private[akka] object MessageSerializer {
       case NonFatal(e) ⇒
         throw new SerializationException(s"Failed to serialize remote message [${message.getClass}] " +
           s"using serializer [${serializer.getClass}].", e)
-    }
+    } finally Serialization.currentTransportInformation.value = oldInfo
   }
 
   def serializeForArtery(serialization: Serialization, outboundEnvelope: OutboundEnvelope, headerBuilder: HeaderBuilder, envelope: EnvelopeBuffer): Unit = {
     val message = outboundEnvelope.message
     val serializer = serialization.findSerializerFor(message)
+    val oldInfo = Serialization.currentTransportInformation.value
+    try {
+      if (oldInfo eq null)
+        Serialization.currentTransportInformation.value = serialization.serializationInformation
+
+      headerBuilder setSerializer serializer.identifier
+      headerBuilder setManifest Serializers.manifestFor(serializer, message)
+      envelope.writeHeader(headerBuilder, outboundEnvelope)
 
-    headerBuilder setSerializer serializer.identifier
-    headerBuilder setManifest Serializers.manifestFor(serializer, message)
-    envelope.writeHeader(headerBuilder, outboundEnvelope)
+      serializer match {
+        case ser: ByteBufferSerializer ⇒ ser.toBinary(message, envelope.byteBuffer)
+        case _                         ⇒ envelope.byteBuffer.put(serializer.toBinary(message))
+      }
 
-    serializer match {
-      case ser: ByteBufferSerializer ⇒ ser.toBinary(message, envelope.byteBuffer)
-      case _                         ⇒ envelope.byteBuffer.put(serializer.toBinary(message))
-    }
+    } finally Serialization.currentTransportInformation.value = oldInfo
```

Comment:
Something I don't quite get here is why we need to do the stash/unstash of the old `currentTransportInformation` value, isn't it always the same in the same actorsystem? Is it to avoid a thread local leak?

Reply:
Leaving the value in the thread is not an option. Bad practice, and the same thread pool can actually be used for several actor systems.